tests/test_cli.py

```diff
@@ -0,0 +1,410 @@
+
+    # the setUp and tearDown methods prevent ResourceWarnings from happening
+    def setUp(self):
+        self.patch_stack = ExitStack().__enter__()
+        # disable Engine, so cli actions don't accidentally start one
+        self.patch_stack.enter_context(patch('apluslms_roman.cli.Engine'))
+
+    def tearDown(self):
+        self.patch_stack.close()
+
+    def command_test(self, *command, config=None, config_fn=None,
+            settings=None, exit_code=0):
+        files = {}
+        args = []
+
+        if config:
+            if not isinstance(config, str):
+                config = json_dumps(config)
+            files[config_fn or CONFIG_FN] = config
+        if config_fn:
+            args.extend(('--file', config_fn))
+        if settings:
+            if not isinstance(settings, str):
+                settings = json_dumps(settings)
+            files[SETTINGS_FN] = settings
+            args.extend(('--config', SETTINGS_FN))
+
+        if len(command) == 1 and isinstance(command[0], str):
+            command = shlex_split(command[0])
+        args.extend(command)
+
+        class CliExit(Exception):
+            pass
+
+        with ExitStack() as ctx:
+            # disable '_build' folder fix
+            ctx.enter_context(patch('apluslms_roman.builder.isdir', return_value=True))
+            # mock cli exit, so SystemExit is not raised and we get return code directly
+            exit_mock = ctx.enter_context(patch('apluslms_roman.cli.exit',
+                side_effect=CliExit))
+            # mock files
+            vfs = ctx.enter_context(mock_files(files))
+            # capture stdio
+            out, err = ctx.enter_context(capture_output())
+
+            try:
+                # call main
+                with self.assertRaises(CliExit, msg="cli.main() didn't call cli.exit()"):
+                    cli.main(args=args)
+
+                # check exit
+                exit_mock.assert_called_once()
+                exit_args = exit_mock.call_args[0]
+                if exit_code is not None:
+                    returned_code = exit_args and exit_args[0] or 0
+                    self.assertEqual(returned_code, exit_code,
+                        msg="cli.main exited with unexpect return code")
+
+                # cli.exit writes message to stderr, but it was mocked..
+                if len(exit_args) == 2:
+                    print(exit_args[1], file=err)
+                return self.CommandResult(out.getvalue(), err.getvalue(), vfs)
+            except Exception as e:
+                if exit_mock.called:
+                    exit_args = exit_mock.call_args[0]
+                    returned_code = exit_args and exit_args[0] or 0
+                else:
+                    returned_code = '?'
+                print("Error traceback:", file=err)
+                print(str(e), file=err)
+                print_tb(e.__traceback__, limit=10, file=err)
```

Comment:
It's bit questionable to print test output to `err`, which is the stderr of the `cli:main` execution. In theory, `raise AssertionError(...) from e` should be enough, but I get that the `unittest` framework seems to eat some of the important traceback lines.

How about something like?:

```
raise AssertionError(
    "cli.main({!r}) -> {} raised an exception\n\n{}\n"
    "Roman output:\n{}\n{}"
    .format(
        args,
        returned_code,
        format_exc(),
        out.getvalue().strip(),
        err.getvalue().strip(),
    ))
```

It seems `format_exc()` renders a string with all of the exception details.

Reply:
The original problem with `unittest` is that it probably uses `print_exc(limit=2)`. The positive limit means, that 2 frames from our current location forward and thus doesn't show the line which actually raised the error. I would presume `limit=-10` would be a better default to avoid this.

The best would be to show 5 frames up from current location and -3 last frames (presuming there is more than 8 frames) for the latest exception and -2 for all previous errors in the chain. Implementing that here would be waste of a time though.

Reply:
If I recall correctly I couldn't find a method for turning the error into a traceback string without printing, and I would've jsut printed the info but the mocked out and err ate it so I ended with that. :p

Reply:
Apparently all I had to do was scroll the docs down a bit... oh well